api/envoy/extensions/filters/http/rbac/v3/rbac.proto

```diff
@@ -18,19 +22,41 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#extension: envoy.filters.http.rbac]
 
 // RBAC filter config.
+// [#next-free-field: 6]
 message RBAC {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.http.rbac.v2.RBAC";
 
   // Specify the RBAC rules to be applied globally.
   // If absent, no enforcing RBAC policy will be applied.
   // If present and empty, DENY.
-  config.rbac.v3.RBAC rules = 1;
+  // If both rules and matcher are configured, rules will be ignored.
+  config.rbac.v3.RBAC rules = 1
+      [(udpa.annotations.field_migrate).oneof_promotion = "rules_specifier"];
+
+  // The match tree to use when resolving RBAC action for incoming requests. Requests do not
+  // match any matcher will be denied.
+  // If absent, no enforcing RBAC matcher will be applied.
+  // If present and empty, deny all requests.
+  xds.type.matcher.v3.Matcher matcher = 4 [
+    (udpa.annotations.field_migrate).oneof_promotion = "rules_specifier",
+    (xds.annotations.v3.field_status).work_in_progress = true
+  ];
 
   // Shadow rules are not enforced by the filter (i.e., returning a 403)
   // but will emit stats and logs and can be used for rule testing.
   // If absent, no shadow RBAC policy will be applied.
-  config.rbac.v3.RBAC shadow_rules = 2;
+  // If both shadow rules and shadow matcher are configured, shadow rules will be ignored.
+  config.rbac.v3.RBAC shadow_rules = 2
+      [(udpa.annotations.field_migrate).oneof_promotion = "shadow_rules_specifier"];
+
+  // The match tree to use for emitting stats and logs which can be used for rule testing for
+  // incoming requests.
+  // If absent, no shadow matcher will be applied.
+  xds.type.matcher.v3.Matcher shadow_matcher = 5 [
```

Comment:
I have a feeling that "shadow actions" would make more sense as part of a single matcher tree. The reason is that it's not clear how to move actions easily between `shadow_matcher` and `matcher` because that requires taking a union of two match trees. The previous iteration this was easier due to it being a list always, so applying a rule amounted to moving an item from one list to another. WDYT?

Reply:
Yes, both shadow policies and shadow matcher are just used for testing.
I am not sure if it is easy to register both `Action` and `ShadowAction`. To me, I think it is good to add a field `generate_stats_only` to `Action` which will permit all matched requests but generate corresponding stats.

Reply:
Add some thoughts. Shadow rules and shadow matcher are for test only, and they may have opposite actions. So maybe it is impossible for us to combine them.

Reply:
I see. Yeah, I think it's possible to have overlaps in the tree and shadow match would prevent exploring nested sub-trees for the other matches. Until we permit multiple matches, it seems like it's up to the user to perform the union of shadow and real matches after testing.